File: julia/06-io-operations/safe_read.jl
filename = "missing.txt"

# Reading a nonexistent file throws a SystemError
try
    data = read(filename, String)
    println(data)
catch err
    if isa(err, SystemError)
        println("Error: could not open '$filename'")
    else
        rethrow(err)   # re-raise anything we didn't expect
    end
end

# Check existence before reading to avoid the error entirely
if isfile(filename)
    println("$filename is $(filesize(filename)) bytes")
else
    println("$filename is not present")
end

File: lua/06-io-operations/safe_read.lua
local filename = "missing.txt"

-- io.open returns nil + message on failure instead of raising
local file, err = io.open(filename, "r")
if file then
    local data = file:read("a")
    file:close()
    print(data)
else
    print("Error: could not open '" .. filename .. "'")
    print("Reason from OS: " .. err)
end

-- assert() promotes the nil+message convention into a raised error,
-- which pcall then catches so the script keeps running
local ok = pcall(function()
    local f = assert(io.open(filename, "r"))
    f:close()
end)
print("pcall succeeded: " .. tostring(ok))
